### ISSUE TITLE
Ng2-smart-table custom Render component

### ISSUE THREAD

**Issue by atlantageek** [akveo/ng2-smart-table#533]
Im trying to use a custom Renderer component in my ng2-smart-table but am getting the following error.

    No component factory found for undefined. Did you add it to @NgModule.entryComponents?
        at…, …}
Whats Strange is that the component says its undefined. 

My component looks like this.

    import { Component, OnInit, Input } from '@angular/core';
    import { ChannelTableModeService } from '../../../../../../services/channel-table-mode.service';
    import { ViewCell } from 'ng2-smart-table';
    
    @Component({
      selector: 'app-channel-attribute-renderer',
      template: '{{renderValue}} --'
    })
    export class ChannelAttributeRendererComponent implements OnInit {
    
    constructor() { }
      renderValue: string;
      @Input() value: string | number;
      @Input() rowData: any;
      ngOnInit() {
         this.renderValue =  '<' + this.value + '>';
      }
    
    }

I've also made sure that the component is in the module.

        @NgModule({
          imports: [
            Ng2SmartTableModule,
            CommonModule,
            ChannelProfileRoutingModule,    
            VeexCommonModule,
            TranslateModule,
            FormsModule
          ],
          entryComponents: [ChannelAttributeRendererComponent, PasswordRenderComponent],
          providers: [ChannelTableModeService],
          declarations: [ PasswordRenderComponent, ChannelProfileComponent, ChannelTableComponent, ServiceTypeComponent, ServiceGroupComponent, ChannelPlanEditorComponent, ChannelAttributeRendererComponent]
        })
    
    export class ChannelProfileModule { }

So does anyone see the issue.  Why does he factory think that the component is undefined.

The configuration form my smart table looks like this:

    this.settings = { 
    columns: {
      'Channel': {title: 'Channel'},
      'Channel Label': {title: 'Channel Name'},
      'Video': {title: 'Video', type: 'custom', renderComponent: 
    PasswordRenderComponent}
    }}

**Comment by balaguero** [comment]
I am having the same issue.

**Comment by trancongduy** [comment]
I am having the same issue.

**Comment by fthsvk** [comment]
Hi how to access rowData in renderComponent?
How can I access each row data?

**Comment by jkon** [comment]
I'm having the same issue as @atlantageek. I have my custom render component and the component trying to use it in both the entryComponents and the declarations, but I still get the no component factory error coming from ViewCell. I'm trying to render a ngbTooltip inside one of my columns

**Comment by Nik1990** [comment]
Hi, i have the some problem @jkon . You resolve it?
Thanks

**Comment by jkon** [comment]
No, I haven’t. I may have to use a different data grid for my project. I also need to find a way to freeze the headers and some columns which I can’t find a good way of doing with this package

**Comment by dandro32** [comment]
I have same issue when I try to use custom editor

**Comment by Nik1990** [comment]
I don't know if this is a case, but i resolved it by:
- removing ng2-smart-table package
- update npm to latest version
- reinstall ng2-smart-table package

**Comment by violetma7** [comment]
I also have this issue. No solution yet.

**Comment by jean-felipe** [comment]
Hey guys, putting the element inside `entryComponents` solved the problem to me:
```js
entryComponents: [CustomRendererComponent]

**Comment by lwalkenshaw** [comment]
Had the same "No component factory found for undefined. Did you add it to @NgModule.entryComponents?"
I fixed it by changing "renderComponent" to "component" in the column definition.

I found this by looking at the error in dev tools, it tries to get the render component from ".component" which is undefined if you use renderComponent

**Comment by IgnacioYanjari** [comment]
> Hey guys, putting the element inside `entryComponents` solved the problem to me:
> 
> ```js
> import { CustomRendererComponent } from './custom-renderer.component';
> entryComponents: [CustomRendererComponent]
> ```

that worked for me, thanks!!

**Comment by jovanica7** [comment]
I have the same issue - for ViewCellComponent. Although I just specified
import { ViewCell } from 'ng2-smart-table'; - not ViewCellComponent. 

Aded this into app.module.ts
import { ViewCellComponent } from 'ng2-smart-table/components/cell/cell-view-mode/view-cell.component';
import { ButtonRenderComponent } from './project/........';

and in the @NgModule
entryComponents: [ViewCellComponent, ButtonRenderComponent]

but same issue.

**Comment by victorlenz** [comment]
> Had the same "No component factory found for undefined. Did you add it to @NgModule.entryComponents?"
> I fixed it by changing "renderComponent" to "component" in the column definition.
> 
> I found this by looking at the error in dev tools, it tries to get the render component from ".component" which is undefined if you use renderComponent

Yes, that does worked for me. Thanks

**Comment by sEpiphane** [comment]
Had the same "No component factory found for undefined. "
isn't working for me

**Comment by sEpiphane** [comment]
> I have the same issue - for ViewCellComponent. Although I just specified
> import { ViewCell } from 'ng2-smart-table'; - not ViewCellComponent.
> 
> Aded this into app.module.ts
> import { ViewCellComponent } from 'ng2-smart-table/components/cell/cell-view-mode/view-cell.component';
> import { ButtonRenderComponent } from './project/........';
> 
> and in the @NgModule
> entryComponents: [ViewCellComponent, ButtonRenderComponent]
> 
> but same issue.

wep bingo ....beautiful

**Comment by JimmyLovely** [comment]
Had the same "No component factory found for undefined. "  in lazy-loaded module

**Comment by JimmyLovely** [comment]
> Had the same "No component factory found for undefined. " in lazy-loaded module

It works for me after upgrading version to 1.5.0

**Comment by csimpf** [comment]
> Had the same "No component factory found for undefined. Did you add it to @NgModule.entryComponents?"
> I fixed it by changing "renderComponent" to "component" in the column definition.
> 
> I found this by looking at the error in dev tools, it tries to get the render component from ".component" which is undefined if you use renderComponent

I fixed mine by doing the opposite! Changed `component` to `renderComponent`.

**Comment by minhtam-nguyennhat** [comment]
still not workk for version 1.5.0

**Comment by zeinebKH** [comment]
hi
i want to add link in cell in this link there is video but the video dosen't work
type: { title: 'Type', valuePrepareFunction: (type) => { if(type==="A"){ return 'Action'; } else { return 'Error'; } } }, videoName: { title: 'video', type: 'custom', renderComponent: LinkVideoComponent, onComponentInitFunction(instance) { instance.save.subscribe(row => { alert(${row.videoName} !!!!) }); } },
LinkVideoComponent:
`export class LinkVideoComponent implements OnInit,ViewCell {

renderValue: string;

@input() value: string | number;
@input() rowData: any;

@output() save: EventEmitter = new EventEmitter();
@ViewChild('videoPlayer') videoplayer: ElementRef;

@ViewChild("content") modalContent: TemplateRef;

constructor(private modalService: NgbModal) {}

ngOnInit() {
this.returnLinkVideo();
}
returnLinkVideo() {
this.renderValue = '' + this.value + '';
}

onClick() {
this.save.emit(this.rowData);
this.modalService.open(this.value, { size: 'lg'});
this.videoplayer.nativeElement.play();

}`
Html
<p #videoplayer (click)="onClick()" >{{ renderValue }}

the video in o folder in the same folder project i didn't know how to get it, any help
